infra_event_notifier/backends/jira.py

```diff
@@ -1,8 +1,240 @@
+    else:
+        _create_jira_issue(
+            jira, fields.title, fields.text, fields.tags, fields.issue_type
+        )
+
+
+def _create_jira_issue(
+    jira: JiraConfig,
+    title: str,
+    body: str,
+    tags: Mapping[str, str],
+    issue_type: str,
+) -> Any | None:
+    """
+    Attempts to create a new jira issue.
+    """
+    api_key = jira.api_key
+    api_url = f"{jira.url}/rest/api/2/issue"
+    payload = {
+        "fields": {
+            "project": {"key": jira.project_key},
+            "summary": title,
+            "description": body,
+            "issuetype": {"name": issue_type},
+            "labels": [f"{k}:{v}" for k, v in tags.items()],
+        }
+    }
+    json_data = json.dumps(payload)
+    data = json_data.encode("utf-8")
+    req = Request(api_url, data=data, method="POST")
+    req = http_basic_auth(jira.user_email, api_key, req)
+    req.add_header("Content-Type", "application/json")
+
+    with urllib.request.urlopen(req) as response:
+        status = response.status
+        if status == 201:
+            return response
+        else:
+            raise JiraApiException(
+                f"Failed to create issue: {response.status}, {response.text}"
+            )
+
+
+def _update_jira_issue(
+    jira: JiraConfig, issue_key: str, body: str
+) -> Any | None:
+    """
+    Attempts to update a jira issue given the issue key.
+    """
+    api_key = jira.api_key
+    api_url = f"{jira.url}/rest/api/2/issue/{issue_key}"
+    payload = {"fields": {"description": body}}
+    json_data = json.dumps(payload)
+    data = json_data.encode("utf-8")
+    req = Request(api_url, data=data, method="PUT")
+    req = http_basic_auth(jira.user_email, api_key, req)
+    req.add_header("Content-Type", "application/json")
+
+    with urllib.request.urlopen(req) as response:
+        status = response.status
+
+        if status == 204:
+            return response
+        else:
+            raise JiraApiException(
+                f"Failed to update issue: "
+                f"{response.status_code}, {response.text}"
+            )
+
+
+def _add_jira_comment(
+    jira: JiraConfig, issue_key: str, comment: str
+) -> Any | None:
+    """
+    Adds a comment to the given jira issue.
+    """
+    api_key = jira.api_key
+    api_url = f"{jira.url}/rest/api/2/issue/{issue_key}/comment"
+    payload = {"body": comment}
+    json_data = json.dumps(payload)
+    data = json_data.encode("utf-8")
+    req = Request(api_url, data=data, method="POST")
+    req = http_basic_auth(jira.user_email, api_key, req)
+    req.add_header("Content-Type", "application/json")
+
+    with urllib.request.urlopen(req) as response:
+        status = response.status
+
+        if status == 201:
+            return response
+        else:
+            raise JiraApiException(
+                f"Failed to update issue: "
+                f"{response.status_code}, {response.text}"
+            )
```

Comment:
The code to make the call to jira and process the response is very similar between the three use case, please consider factor it into a function as only URL, payload and response code are different.